Tensors/src/main/scala/com/thoughtworks/compute/Tensors.scala

```
@@ -1125,14 +1125,55 @@ trait Tensors extends OpenCL {
     }
 
     /** Returns an asynchronous task to read this [[Tensor]] into a [[scala.Array]],
-      * which is linearized in row-majoy order.
+      * which is linearized in row-major order.
       *
       * @group slow
       */
     def flatArray: Future[Array[closure.JvmValue]] = {
       flatBuffer.intransitiveMap(closure.valueType.memory.toArray).run
     }
 
+    /** Recursive function to reshape a flat array into a multi-dimensional one
+      */
+    private[Tensors] def reshapeArray(array:Array[_], shape:Array[Int]): Array[_] = {
+      if (shape.product != array.length) {
+        throw new IllegalArgumentException
+      }
+      if (shape.length < 2) {
+        array
+      } else {
+        val lastDimLength = shape(shape.length - 1)
+        val oneReduced = Array.tabulate(shape.product/lastDimLength)(i => array.slice(i*lastDimLength,(i+1)*lastDimLength))
+        reshapeArray(oneReduced, shape.slice(0, shape.length-1))
+      }
+    }
+
+    /** Recursive function to reshape a flat Seq into a multi-dimensional one
+      */
+    private[Tensors] def reshapeSeq(a:Seq[_], b:Array[Int]): Seq[_] = {
```

When `shape` is `Array.empty`, the tensor is a scalar value, not a `Seq`.

This method should either throw an exception for scalar value, or change the return type.